File: app/src/main/java/com/nacare/capture/ui/events/EventAdapter.java
package com.nacare.capture.ui.events;

import static com.nacare.capture.data.service.StyleBinderHelper.setBackgroundColor;
import static com.nacare.capture.data.service.StyleBinderHelper.setState;

import android.graphics.Color;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import androidx.paging.DataSource;
import androidx.paging.PagedListAdapter;

import com.nacare.capture.R;
import com.nacare.capture.data.Sdk;
import com.nacare.capture.data.service.ActivityStarter;
import com.nacare.capture.data.service.DateFormatHelper;
import com.nacare.capture.ui.base.DiffByIdItemCallback;
import com.nacare.capture.ui.base.ListItemWithSyncHolder;
import com.nacare.capture.ui.event_form.EventFormActivity;
import com.nacare.capture.ui.tracker_import_conflicts.TrackerImportConflictsAdapter;

import org.hisp.dhis.android.core.category.CategoryOptionCombo;
import org.hisp.dhis.android.core.event.Event;
import org.hisp.dhis.android.core.maintenance.D2Error;
import org.hisp.dhis.android.core.organisationunit.OrganisationUnit;
import org.hisp.dhis.android.core.program.ProgramStageDataElement;
import org.hisp.dhis.android.core.trackedentity.TrackedEntityDataValue;

import java.util.ArrayList;
import java.util.List;

public class EventAdapter extends PagedListAdapter<Event, ListItemWithSyncHolder> {

    private final AppCompatActivity activity;
    private DataSource<?, Event> source;

    public EventAdapter(AppCompatActivity activity) {
        super(new DiffByIdItemCallback<>());
        this.activity = activity;
    }

    @NonNull
    @Override
    public ListItemWithSyncHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.list_item_update, parent, false);
        return new ListItemWithSyncHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull ListItemWithSyncHolder holder, int position) {
        Event event = getItem(position);
        List<TrackedEntityDataValue> values = new ArrayList<>(event.trackedEntityDataValues());
        holder.title.setText(orgUnit(event.organisationUnit()).displayName());
        holder.subtitle1.setText(valueAt(values, event.programStage()));
        holder.subtitle2.setText(optionCombo(event.attributeOptionCombo()).displayName());
        holder.rightText.setText(DateFormatHelper.formatDate(event.eventDate()));


        holder.dateTextView.setText(DateFormatHelper.formatSimpleDate(event.eventDate()));
        holder.statusTextView.setText(event.status().name());
        holder.actionTextView.setText("Edit");
        int colorBlack = ContextCompat.getColor(holder.itemView.getContext(), R.color.black);

        holder.dateTextView.setTextColor(colorBlack);
        holder.statusTextView.setTextColor(colorBlack);
        holder.actionTextView.setTextColor(colorBlack);
        holder.icon.setImageResource(R.drawable.ic_programs_black_24dp);
        holder.delete.setVisibility(View.VISIBLE);
        holder.delete.setOnClickListener(view -> {
            try {
                Sdk.d2().eventModule().events().uid(event.uid()).blockingDelete();
                invalidateSource();
                notifyDataSetChanged();
            } catch (D2Error d2Error) {
                d2Error.printStackTrace();
            }
        });
        setBackgroundColor(R.color.colorAccentDark, holder.icon);
        setState(event.aggregatedSyncState(), holder.syncIcon);
        setConflicts(event.uid(), holder);

        holder.itemView.setOnClickListener(view -> ActivityStarter.startActivity(
                activity,
                EventFormActivity.getFormActivityIntent(
                        activity,
                        event.uid(),
                        event.program(),
                        event.organisationUnit(),
                        EventFormActivity.FormType.CHECK
                ), false
        ));
    }

    private OrganisationUnit orgUnit(String orgUnitUid) {
        return Sdk.d2().organisationUnitModule().organisationUnits().uid(orgUnitUid).blockingGet();
    }

    private String valueAt(List<TrackedEntityDataValue> values, String stageUid) {
        for (TrackedEntityDataValue dataValue : values) {
            ProgramStageDataElement programStageDataElement = Sdk.d2().programModule().programStageDataElements()
                    .byDataElement().eq(dataValue.dataElement())
                    .byProgramStage().eq(stageUid)
                    .one().blockingGet();
            if (programStageDataElement.displayInReports()) {
                return String.format("%s: %s", programStageDataElement.displayName(), dataValue.value());
            }
        }

        return null;
    }

    private CategoryOptionCombo optionCombo(String attrOptionCombo) {
        return Sdk.d2().categoryModule().categoryOptionCombos().uid(attrOptionCombo).blockingGet();
    }

    private void setConflicts(String trackedEntityInstanceUid, ListItemWithSyncHolder holder) {
        TrackerImportConflictsAdapter adapter = new TrackerImportConflictsAdapter();
        holder.recyclerView.setAdapter(adapter);
        adapter.setTrackerImportConflicts(Sdk.d2().importModule().trackerImportConflicts()
                .byTrackedEntityInstanceUid().eq(trackedEntityInstanceUid).blockingGet());
    }

    public void setSource(DataSource<?, Event> dataSource) {
        this.source = dataSource;
    }

    public void invalidateSource() {
        source.invalidate();
    }
}

File: app/src/main/java/com/intellisoft/nacare/adapter/EventAdapter.kt
package com.intellisoft.nacare.adapter

import android.content.Context
import android.graphics.PorterDuff
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.card.MaterialCardView
import com.intellisoft.nacare.helper_class.FormatterClass
import com.intellisoft.nacare.helper_class.SubmissionsStatus
import com.intellisoft.nacare.room.EventData
import com.nacare.capture.R


class EventAdapter(
    private var dataList: List<EventData>,
    private val context: Context,
    private val click: (EventData) -> Unit,
    private val syncEvent: (EventData) -> Unit
) : RecyclerView.Adapter<EventAdapter.Pager2ViewHolder>() {

    inner class Pager2ViewHolder(itemView: View) :
        RecyclerView.ViewHolder(itemView), View.OnClickListener {
        val formatterClass = FormatterClass()
        val syncIcon: ImageView = itemView.findViewById(R.id.sync_icon)
        val eventStatus: ImageView = itemView.findViewById(R.id.eventStatus)
        val eventDate: TextView = itemView.findViewById(R.id.event_date)
        val organisationUnit: TextView = itemView.findViewById(R.id.organisationUnit)
        val eventInfo: TextView = itemView.findViewById(R.id.eventInfo)
        private val eventCard: MaterialCardView = itemView.findViewById(R.id.eventCard)

        init {
            eventCard.setOnClickListener(this)
        }

        override fun onClick(p0: View?) {
            val patient = dataList[adapterPosition]
            click(patient)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): Pager2ViewHolder {
        return Pager2ViewHolder(
            LayoutInflater.from(parent.context).inflate(R.layout.item_patient, parent, false)
        )
    }

    override fun onBindViewHolder(holder: Pager2ViewHolder, position: Int) {
        val data = dataList[position]
        val date = data.date
        val org = data.orgUnitName
        holder.organisationUnit.text = org
        holder.eventDate.text = date
        when (data.status) {
            SubmissionsStatus.DRAFT.name -> {
                holder.eventStatus.setImageResource(R.drawable.ic_event_status_open)
                holder.syncIcon.visibility = View.GONE
            }

            "completed" -> {
                holder.eventStatus.setImageResource(R.drawable.completeddoc)
                holder.eventStatus.setColorFilter(
                    ContextCompat.getColor(context, R.color.green), PorterDuff.Mode.SRC_IN
                )
                if (!data.synced) {
                    holder.syncIcon.visibility = View.VISIBLE
                    holder.syncIcon.setColorFilter(
                        ContextCompat.getColor(context, R.color.primary), PorterDuff.Mode.SRC_IN
                    )

                    holder.syncIcon.apply {
                        setOnClickListener {
                            syncEvent(data)
                        }
                    }
                }

            }

            SubmissionsStatus.DUPLICATED.name -> {

            }
        }

    }

    override fun getItemCount(): Int {
        return dataList.size
    }
}